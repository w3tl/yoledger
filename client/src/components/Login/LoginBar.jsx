import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { Menu } from 'semantic-ui-react';

function ProfileHeader({ signout }) {
  return (
    <Fragment>
      <Menu.Item name="profile" as={Link} to="/profile" />
      <Menu.Item name="signout" onClick={signout} />
    </Fragment>
  );
}

ProfileHeader.propTypes = {
  signout: PropTypes.func.isRequired,
};

function LoginBar({ signout, loggedIn }) {
  return (
    <Menu.Menu position="right">
      {loggedIn && <ProfileHeader signout={signout} />}
      {!loggedIn && (
        <Menu.Item as={Link} to="/signin" name="Sign-in" />)}
    </Menu.Menu>
  );
}

LoginBar.propTypes = {
  signout: PropTypes.func.isRequired,
  loggedIn: PropTypes.bool.isRequired,
};

export default LoginBar;
