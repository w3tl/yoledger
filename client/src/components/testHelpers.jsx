import React from 'react';
import { shallow, mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

export const withProvider = Component => props => (
  <MockedProvider {...props}>
    <Component />
  </MockedProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return fixedDate;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
