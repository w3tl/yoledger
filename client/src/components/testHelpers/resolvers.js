import dateResolver from '../../../../server/src/graphql/date';
import assets from '../Asset/mockData';
import expenses from '../Expense/mockData';
import incomes from '../Income/mockData';
import { accounts, periods, budgets } from '../Budget/mockData';
import transactions from '../Transaction/mockData';

const accountsArray = assets.concat(expenses).concat(incomes);

export default {
  Date: dateResolver,
  Query: {
    account(root, { id }) {
      if (id === '666') return { name: 'new account' };
      return accountsArray.find(a => a.id === id);
    },
    accounts(root, { type }) {
      if (type === 'ASSET') return assets;
      if (type === 'EXPENSE') return expenses;
      if (type === 'INCOME') return incomes;
      return [];
    },
    budgets() {
      return { accounts, periods: periods.map(dateStr => new Date(dateStr)) };
    },
    budget() {
      return budgets[0];
    },
    transactions() {
      return transactions.map(({ date, ...other }) => ({ date: new Date(date), ...other }));
    },
    transaction(root, { id }) {
      if (id === '666') return { name: 'new account' };
      const { date, ...other } = transactions.find(a => a.id === id);
      return { date: new Date(date), ...other };
    },
  },
  Mutation: {
    upsertBudget(root, { input: { date } }) {
      return {
        success: true,
        allocation: {
          id: date.toISOString().concat('new account'),
          account: { id: '99', name: 'new account' },
        },
      };
    },
    addAccount(root, { input }) {
      return {
        account: { id: '666', name: input.name },
      };
    },
    addTransaction(root, { input }) {
      return {
        transaction: {
          id: '666',
          ...input,
          source: accountsArray.find(a => a.name === input.source),
          destination: accountsArray.find(a => a.name === input.destination),
        },
      };
    },
    deleteTransaction(root, { id }) {
      return {
        success: true,
        id,
      };
    },
    updateTransaction(root, { id, input }) {
      return {
        transaction: {
          ...transactions.find(t => t.id === id),
          id: '11',
          ...input,
          source: accountsArray.find(a => a.name === input.source),
          destination: accountsArray.find(a => a.name === input.destination),
        },
      };
    },
    signin(root, { login }) {
      return {
        token: `{\\"userId\\":\\"${login}\\"};secret;{\\"expiresIn\\":10}`,
      };
    },
    signout() {
      return {
        success: true,
      };
    },
    changePassword() {
      return {
        token: `{\\"userId\\":\\"${'admin'}\\"};secret;{\\"expiresIn\\":10}`,
      };
    },
  },
};
