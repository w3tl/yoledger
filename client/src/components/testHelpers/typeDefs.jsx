export default `
scalar Date
type Query {
  account(id: ID!): Account
  accounts(type: AccountType!): [Account]

  budgets(dateStart: Date!, count: Int = 5): Budget
  budget(date: Date!): [Allocation]

  transactions(dateStart: Date!, dateEnd: Date, page: Int = 0, itemsPerPage: Int = 20): [Transaction]
  transaction(id: ID!): Transaction
}
type Mutation {
  addAccount(input: AddAccountInput!): AddAccountPayload
  upsertBudget(input: UpsertBudgetInput!): UpsertBudgetPayload
  addTransaction(input: AddTransactionInput!): AddTransactionPayload
  deleteTransaction(id: ID!): DeleteTransactionPayload
  updateTransaction(id: ID!, input: UpdateTransactionInput!): UpdateTransactionPayload
}

type Account {
  id: ID!
  name: String!
  balance: Float
  type: AccountType!
}

type Budget {
  periods: [Date]
  accounts: [Account]
}

type Allocation {
  id: String!
  account: Account!
  amount: Float
  balance: Float
}

enum AccountType {
  ASSET
  EXPENSE
  INCOME
  VENDOR
}

input AddAccountInput {
  name: String!
  balance: Float
  type: AccountType!
}

type AddAccountPayload {
  account: Account
}

input UpsertBudgetInput {
  account: String!
  date: Date!
  amount: Float!
}

type UpsertBudgetPayload {
  success: Boolean
  allocation: Allocation
}

type Transaction {
  id: ID!
  amount: Float!
  source: Account!
  destination: Account!
  date: Date!
}
input AddTransactionInput {
  source: String!
  destination: String!
  amount: Float!
  date: Date
}
type AddTransactionPayload {
  transaction: Transaction
}
type DeleteTransactionPayload {
  success: Boolean!
  id: ID
}
input UpdateTransactionInput {
  amount: Float
  source: String
  destination: String
  date: Date
}
type UpdateTransactionPayload {
  transaction: Transaction
}
`;
