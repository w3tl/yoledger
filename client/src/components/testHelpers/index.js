import React from 'react';
import { shallow, mount } from 'enzyme';
import { ApolloProvider } from 'react-apollo';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { makeExecutableSchema } from 'graphql-tools';
import typeDefs from './typeDefs';
import resolvers from './resolvers';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

const executableSchema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

const cache = new InMemoryCache();
const link = new SchemaLink({ schema: executableSchema });
export const client = new ApolloClient({
  link,
  cache,
});

export const withProvider = Component => props => (
  <ApolloProvider client={client}>
    <Component client={client} />
  </ApolloProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  const date = new Date(fixedDate);
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return date;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
