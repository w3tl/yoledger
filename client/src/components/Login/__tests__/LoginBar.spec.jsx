import React from 'react';
import { mount } from 'enzyme';
import LoginBar from '../LoginBar';
import { testRenderWithoutError } from '../../testHelpers';

jest.mock('../Login', () => () => (
  <div>Login</div>));

jest.mock('react-router-dom', () => ({ // eslint-disable-next-line react/prop-types
  Link: ({ to, children }) => <a href={to}>{children}</a>,
}));

describe('LoginBar', () => {
  testRenderWithoutError(<LoginBar signin={() => {}} signout={() => {}} loggedIn />);

  it('should render signout button when user logged in', () => {
    const signout = jest.fn();
    const output = mount(
      <LoginBar signin={() => {}} signout={signout} loggedIn />,
    );
    expect(output).toMatchSnapshot();
    output.find('Button').simulate('click');
    expect(signout).toHaveBeenCalled();
  });

  it('should render Login component when user not are logged in', () => {
    const output = mount(
      <LoginBar signout={() => {}} signin={() => {}} loggedIn={false} />,
    );
    expect(output).toMatchSnapshot();
  });
});
