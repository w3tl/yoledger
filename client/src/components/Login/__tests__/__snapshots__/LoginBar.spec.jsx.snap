// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginBar renders without error 1`] = `
<div>
  <ProfileHeader
    signout={[Function]}
  />
</div>
`;

exports[`LoginBar should render Login component when user not are logged in 1`] = `
<LoginBar
  loggedIn={false}
  signin={[Function]}
  signout={[Function]}
>
  <div>
    <Component
      signin={[Function]}
    >
      <div>
        Login
      </div>
    </Component>
  </div>
</LoginBar>
`;

exports[`LoginBar should render signout button when user logged in 1`] = `
<LoginBar
  loggedIn={true}
  signin={[Function]}
  signout={[Function]}
>
  <div>
    <ProfileHeader
      signout={[Function]}
    >
      <div>
        <Link
          to="/profile"
        >
          <a
            href="/profile"
          >
            Profile
          </a>
        </Link>
        <Button
          onClick={[Function]}
          type="button"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            Signout
          </button>
        </Button>
      </div>
    </ProfileHeader>
  </div>
</LoginBar>
`;
